liquidctl/driver/nzxt_smart_device.py

```diff
@@ -182,66 +276,244 @@ def get_status(self, **kwargs):
+                        status.append(('Fan {} duty'.format(i + 1), msg[duty_offset + i], '%'))
+                    rpm_offset += 2
+                status.append(('Noise level', msg[noise_offset], 'dB'))
+                num_valid_replies_recvd += 1
+                msg_6702_reply = True
+                continue
+            if not msg_6704_reply and msg[0] == 0x67 and msg[1] == 0x04:
+            # PLACE HOLDER FOR FUTURE WORK
+            # Interpretation of this reply is pending
+                #status.append((' Unknown:', 'status', 'pending'))
+                num_valid_replies_recvd += 1
+                msg_6704_reply = True
+                continue
+        self.device.release()
+        return sorted(status)
+
+    def _write_colors(self, cid, mode, colors, sval):
+        mval, mod3, mod4, mincolors, maxcolors = self._COLOR_MODES[mode]
+        color_count = len(colors)
+        if maxcolors == 40:
+            channel_mod = [0x1E, 0x20][cid] # the purpose of this is unknown, but is based on cmd issued by CAM software
+            led_padding = [0xFF, 0xFF, 0xFF]*(maxcolors - color_count)  # set all remaining LEDs to White
+            leds = list(itertools.chain(*colors)) + led_padding
+            self._write([0x22, 0x10, cid+1, 0x00] + leds[0:60]) # send first 20 colors to device (3 bytes per color)
+            self._write([0x22, 0x11, cid+1, 0x00] + leds[60:])  # send remaining colors to device
+            msg = self.device.read(self._READ_LENGTH) # wait for one reply before issuing command to specify color mode
+            LOGGER.debug('received %s', ' '.join(format(i, '02x') for i in msg))
+            self._write([0x22, 0xA0, cid+1, 0x00, mval, mod3, 0x00, channel_mod, 0x00,
+                         0x00, 0x64, 0x00, 0x32, 0x00, 0x00, 0x01])
+        elif mode == 'wings':  # wings requires special handling
+            for [g, r, b] in colors:
+                self._write([0x22, 0x10, cid+1])  # clear out all independent LEDs
+                self._write([0x22, 0x11, cid+1])  # clear out all independent LEDs
+                color_list = [g, r, b] * 8
+                r2 = (int)(r // 2.5)
+                g2 = (int)(g // 2.5)
+                b2 = (int)(b // 2.5)
+                color_list2 = [g2, r2, b2] * 8
+                r3 = (int)(r2 // 4)
+                g3 = (int)(g2 // 4)
+                b3 = (int)(b2 // 4)
+                color_list3 = [g3, r3, b3] * 8
+                for i in range(8):   #  send color scheme first, before enabling wings mode
+                    if i == 3 or i == 7:
+                        mod = 0x05
+                    else:
+                        mod = 0x01
+                    msg = ([0x22, 0x20, cid+1, i, 0x04, 0x39, 0x00, mod,
+                            0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x06,
+                            0x05, 0x85, 0x05, 0x85, 0x05, 0x85, 0x00, 0x00,
+                            0x00, 0x00, 0x00, 0x00])
```

Comment:
Is this constant suffix entirely necessary?

Reply:
This is a bit tricky. CAM uses 2 different byte streams for my particular setup.

On my `led1` channel with just two 200mm Underglows, it sends 6 colors, hence 0x06 in the 8th byte. This is followed by 6 pairs of `0x05\0x85`. 

But on my `led2` channel with five Aer RGB 2 fans, it sends 8 colors with 0x08 in the 8th byte, followed by 8 pairs of `0x04\0x84`.

So it seems to be generating a command sequence that is proportional to the number of LEDs on a channel??

Reply:
Hopefully it's just proportional to the number of LEDs, and not dependent on the specific device types.

Still, this could be related with being able to mix different HUE 2 accessories in a single channel: handling the differences between accessories could have been delegated to software.  I mean, we're already interpolating the colors for the wings mode.

---

Even so... maybe it works with all zeros?  :roll_eyes:

Reply:
Just tried it and it just produces a single movement from one side to the other. The `wings` movement splits the LEDs in half and the movement begins on the edges and converges to the center. The word "wings" implies two flaps or two halves with a symmetry running down the middle. So those extra bytes are needed, but later I can see what happens with different values.